### PR TITLE
Hamt implementation

### LINKED ISSUE #187: Implement HAMT structure
I believe the rationale is to have the data cbor serializable and be able to generate a consistent CID, confirm @dutterbutter @ec2 @ansermino ?

### PR DESCRIPTION
**Pull request by austinabell**
**Summary of changes**
Changes introduced in this pull request:
- Implements HAMT, I go into a bit of detail in docs of this crate so won't duplicate
  - Tries to match the rust HashMap API as @dignifiedquire had started, but for our use case and to limit database writes, the interface is changed to match the spec closer and only persist the nodes when the Hamt is flushed. Also we need the error Results which get discarded when matching the rust HashMap interface.
  - I do want to clean up a few more things so I will likely come back to this later, as well as benchmarking, but probably after the other milestones are hit

- Updates CI to run all features (test was added in this crate which overrides the hashing function for canonical tree tests so just to make sure this doesn't get broken in future)

This is a big PR so please just comment or ping me to ask about anything, I can't go into that in depth detail now.

**Reference issue to close (if applicable)**

Closes #187

**Other information and links**

### CODE REVIEW

File: ipld/hamt/src/bitfield.rs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    ser::{Serialize, Serializer},
+    serde_bytes,
+};
+use std::u64;
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Bitfield([u64; 4]);
+
+impl Serialize for Bitfield {
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
```
Comment (ec2) [inline ipld/hamt/src/bitfield.rs]: Can just be `Result` instead of `std::result::Result`

File: ipld/hamt/src/bitfield.rs [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+            }
+        }
+
+        let byte_buf = serde_bytes::Bytes::new(&[]);
+        byte_buf.serialize(serializer)
+    }
+}
+
+impl<'de> Deserialize<'de> for Bitfield {
+    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
```
Comment (ec2) [inline ipld/hamt/src/bitfield.rs]: Same here

File: ipld/hamt/src/hamt.rs [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+impl<'a, K: PartialEq, V: PartialEq, S: BlockStore> PartialEq for Hamt<'a, K, V, S> {
+    fn eq(&self, other: &Self) -> bool {
+        self.root == other.root
+    }
+}
+
+impl<'a, K, V, S> Hamt<'a, K, V, S>
+where
+    K: Hash + Eq + std::cmp::PartialOrd + Serialize + DeserializeOwned + Clone,
```
Comment (ec2) [inline ipld/hamt/src/hamt.rs]: Just PartialOrd is fine

File: ipld/hamt/src/node.rs [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+        Node {
+            bitfield: Bitfield::zero(),
+            pointers: Vec::new(),
+        }
+    }
+}
+
+impl<K, V> Node<K, V>
+where
+    K: Hash + Eq + std::cmp::PartialOrd + Serialize + DeserializeOwned + Clone,
```
Comment (ec2) [inline ipld/hamt/src/node.rs]: PartialOrd

File: ipld/hamt/src/pointer.rs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+                            }
+                        } else {
+                            return Ok(());
+                        }
+                    }
+                    // Replace link node with child values
+                    *self = Pointer::Values(child_vals);
+                    Ok(())
+                }
+                _ => Ok(()),
```
Comment (ec2) [inline ipld/hamt/src/pointer.rs]: looks like this should be unreachable too,right?

Comment (austinabell) [reply ipld/hamt/src/pointer.rs]: resolved in person

File: ipld/hamt/src/node.rs [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+                let np = KeyValuePair::new(key, value);
+                vals.insert(idx, np);
+
+                Ok(None)
+            }
+        }
+    }
+
+    /// Internal method to delete entries.
+    pub fn rm_value<Q: ?Sized, S: BlockStore>(
```
Comment (dutterbutter) [inline ipld/hamt/src/node.rs]: I don't think this needs to be public.

File: ipld/hamt/src/node.rs [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+                child.clean()?;
+                Ok(deleted)
+            }
+            Pointer::Values(vals) => {
+                // Delete value
+                for (i, p) in vals.iter().enumerate() {
+                    if key.eq(p.key().borrow()) {
+                        let old = if vals.len() == 1 {
+                            if let Pointer::Values(new_v) = self.rm_child(cindex, idx) {
+                                new_v.into_iter().nth(0).unwrap()
```
Comment (dutterbutter) [inline ipld/hamt/src/node.rs]: Not sure why it's not showing up in our CI check, however I am getting a Clippy error on this line: 

```
error: called `.nth(0)` on a `std::iter::Iterator`
   --> ipld/hamt/src/node.rs:330:33
    |
330 | ...                   new_v.into_iter().nth(0).unwrap()
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^ help: try calling: `new_v.into_iter().next()`
    |
    = note: `-D clippy::iter-nth-zero` implied by `-D warnings`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#iter_nth_zero
```

File: ipld/hamt/src/node.rs [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+        if !self.bitfield.test_bit(idx) {
+            return Ok(None);
+        }
+
+        let cindex = self.index_for_bit_pos(idx);
+        let child = self.get_child(cindex);
+        match child {
+            Pointer::Link(cid) => match store.get::<Node<K, V>>(cid)? {
+                Some(node) => Ok(node.get_value(hashed_key, bit_width, depth + 1, key, store)?),
+                None => Err(Error::Custom("node not found")),
```
Comment (dutterbutter) [inline ipld/hamt/src/node.rs]: Nit: capitalize (N)ode for consistency

File: ipld/hamt/src/node.rs [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+        match child {
+            Pointer::Link(cid) => match store.get::<Node<K, V>>(cid)? {
+                Some(mut node) => {
+                    // Pull value from store and update to cached node
+                    let v =
+                        node.modify_value(hashed_key, bit_width, depth + 1, key, value, store)?;
+                    *child = Pointer::Cache(Box::new(node));
+                    Ok(v)
+                }
+                None => Err(Error::Custom("node not found")),
```
Comment (dutterbutter) [inline ipld/hamt/src/node.rs]: Nit: same

File: ipld/hamt/src/node.rs [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+                Some(mut node) => {
+                    // Pull value from store and update to cached node
+                    let del = node.rm_value(hashed_key, bit_width, depth + 1, key, store)?;
+                    *child = Pointer::Cache(Box::new(node));
+
+                    // Clean to retrieve canonical form
+                    child.clean()?;
+                    Ok(del)
+                }
+                None => Err(Error::Custom("node not found")),
```
Comment (dutterbutter) [inline ipld/hamt/src/node.rs]: Nit: same

File: ipld/hamt/src/node.rs (hunk 7 above)
Comment (austinabell) [reply ipld/hamt/src/node.rs]: from updated rust toolchain I'm seeing it now, nice catch

**Pull request merged**